### PR TITLE
Pg upgrade database in tablespace

### PR DESCRIPTION
**Pull request by berlin-ab**
Add gpdb5 tablespace paths to the DbInfo struct via the old tablespaces file.

pg_upgrade assumes it is running in `segment` mode unless specified otherwise. All of our regress tests are running pg_upgrade against the master, so I needed to add the `--mode=dispatcher` to get the tests to pass.

### CODE REVIEW

File: contrib/pg_upgrade/test/integration/scenarios/filespaces_to_tablespaces.c [hunk 1]
```diff
@@ -63,6 +63,57 @@ aFilespaceExistsInTheFiveClusterWithATableAndData(void)
 	PQfinish(connection5);
 }
 
+static void
+aDatabaseInAFilespaceExistsInTheFiveClusterWithATableAndData(void)
```
Comment (soumyadeep2007) [inline contrib/pg_upgrade/test/integration/scenarios/filespaces_to_tablespaces.c]: I was wondering about the case where we have the following:
```
CREATE DATABASE db1 TABLESPACE foo;
\c db1
CREATE TABLE tab1 TABLESPACE bar; -- table in database but in different tablespace
```
Maybe it should be in a different PR? I'm okay either way. I think it will be a common scenario for upgrade.

File: contrib/pg_upgrade/info.c [hunk 2]
```diff
@@ -378,8 +434,14 @@ get_db_infos(ClusterInfo *cluster)
...
 
+	/*
+	 * greenplum specific indexes
+	 */
+	int         i_tablespace_oid;
+	int         i_is_default_tablespace;
+
 	snprintf(query, sizeof(query),
-			 "SELECT d.oid, d.datname, %s "
+			 "SELECT d.oid, d.datname, t.oid as tablespace_oid, (t.spcname IN ('pg_default', 'pg_global'))::int as is_default_tablespace, %s "
```
[the remark this answers is not in the document]
Comment (berlin-ab) [reply contrib/pg_upgrade/info.c]: Great suggestion.

File: contrib/pg_upgrade/test/integration/scenarios/filespaces_to_tablespaces.c (hunk 1 above)
Comment (berlin-ab) [reply contrib/pg_upgrade/test/integration/scenarios/filespaces_to_tablespaces.c]: Good catch. I'll start a separate thread for that.

**Comment by berlin-ab** [comment]
> Looks good to me. One or two tweaks needed.
> Just curious: could we have tested these with the regress style or combined with the regress style?

I considered doing a regress test, but I kept it next to the other tablespace test for consistency.

File: contrib/pg_upgrade/info.c [hunk 3]
```diff
@@ -358,6 +358,62 @@ get_db_and_rel_infos(ClusterInfo *cluster)
...
+ */
+static char *
+determine_db_tablespace_path(ClusterInfo *currentCluster,
+                             char *spclocation,
+                             Oid tablespace_oid,
+                             int is_default_tablespace)
+{
+    if (currentCluster != &old_cluster
+            || !is_gpdb_version_with_filespaces(currentCluster)
+            || is_default_tablespace)
```
Comment (dgkimura) [inline contrib/pg_upgrade/info.c]: I follow the last 2 conditions based on the function comments, but I didn't follow why is the first condition `currentCluster != &old_cluster` needed?

File: contrib/pg_upgrade/info.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (berlin-ab) [reply contrib/pg_upgrade/info.c]: Done.

File: contrib/pg_upgrade/tablespace_gp.c [hunk 4]
```diff
@@ -13,72 +13,80 @@
 static char *
 get_generated_old_tablespaces_file_path(void)
 {
-	char current_working_directory[MAXPGPATH];
+    char current_working_directory[MAXPGPATH];
```
[the remark this answers is not in the document]
Comment (berlin-ab) [reply contrib/pg_upgrade/tablespace_gp.c]: Good point. Sorry I missed this earlier.

[the remark this answers is not in the document]
Comment (berlin-ab) [reply contrib/pg_upgrade/tablespace_gp.c]: I pushed another commit.

File: contrib/pg_upgrade/info.c [hunk 5]
```diff
@@ -695,14 +740,13 @@ get_rel_infos(ClusterInfo *cluster, DbInfo *dbinfo)
...
 		{
-			bool is_old_cluster = old_cluster.major_version == cluster->major_version;
-
-			if (is_old_cluster && is_gpdb_version_with_filespaces(&old_cluster))
-			{
-				tablespace = get_tablespace_path_for_old_cluster(
-					old_cluster.old_tablespace_file_contents,
-					tablespace_oid);
-			}
+            if (cluster == &old_cluster
```
[the remark this answers is not in the document]
Comment (berlin-ab) [reply contrib/pg_upgrade/info.c]: It's more exact.

File: contrib/pg_upgrade/test/integration/scenarios/filespaces_to_tablespaces.c [hunk 6]
```diff
@@ -151,6 +194,28 @@ aTablespaceShouldHaveBeenCreatedOnSixCluster(void)
 	stopGpdbSixCluster();
 }
 
+static void
+theDatabaseShouldBeInTheTablespaceOnTheNewCluster(void)
```
[the remark this answers is not in the document]
Comment (berlin-ab) [reply contrib/pg_upgrade/test/integration/scenarios/filespaces_to_tablespaces.c]: I think that we can trust the DDL statement to work from the perspective of upgrade.

[the remark this answers is not in the document]
Comment (berlin-ab) [reply contrib/pg_upgrade/test/integration/scenarios/filespaces_to_tablespaces.c]: I guess theres an argument that we should check that pg_dump has the correct behavior, but I think we could add that test to pg_dump to have the test at the right level of abstraction.
